src/azul/maintenance.py

```diff
@@ -0,0 +1,355 @@
+        ]
+
+    def _now(self):
+        return datetime.now(UTC)
+
+    def add_event(self, event: MaintenanceEvent):
+        """
+        Add the given event to this schedule unless doing so would invalidate
+        this schedule.
+        """
+        events = self.events
+        try:
+            self.events = events.copy()
+            self.events.append(event)
+            self.events.sort(key=attrgetter('start'))
+            self.validate()
+        except BaseException:
+            self.events = events
+            raise
+
+    def adjust_event(self, additional_duration: timedelta):
+        event = self.active_event()
+        reject(event is None, 'No active event')
+        event.planned_duration += additional_duration
+        self._heal(event, iter(self.pending_events()))
+        assert self.active_event() is not None
+        return event
+
+    def cancel_event(self, index: int) -> MaintenanceEvent:
+        event = self.pending_events()[index]
+        self.events.remove(event)
+        self.validate()
+        return event
+
+    def start_event(self) -> MaintenanceEvent:
+        pending = iter(self.pending_events())
+        event = next(pending, None)
+        reject(event is None, 'No events pending to be started')
+        event.actual_start = self._now()
+        self._heal(event, pending)
+        assert event == self.active_event()
+        return event
+
+    def end_event(self) -> MaintenanceEvent:
+        event = self.active_event()
+        reject(event is None, 'No active event')
+        event.actual_end = self._now()
+        self._heal(event, iter(self.pending_events()))
+        assert self.active_event() is None
+        return event
+
+    def _heal(self,
+              event: MaintenanceEvent,
+              pending: Iterator[MaintenanceEvent]):
+        for next_event in pending:
+            if next_event.planned_start < event.end:
+                next_event.planned_start = event.end
+            event = next_event
+        self.validate()
+
+
+class MaintenanceService:
+
+    @property
+    def bucket(self):
+        return aws.shared_bucket
+
+    @property
+    def object_key(self):
+        return f'azul/{config.deployment_stage}/azul.json'
+
+    @cached_property
+    def client(self):
+        return aws.s3
+
+    @property
+    def _get_schedule(self) -> JSON:
+        try:
+            response = self.client.get_object(Bucket=self.bucket,
+                                              Key=self.object_key)
+        except self.client.exceptions.NoSuchKey:
+            raise StorageObjectNotFound
+        else:
+            return json.loads(response['Body'].read())
+
+    @property
+    def get_schedule(self) -> MaintenanceSchedule:
+        schedule = self._get_schedule
+        schedule = MaintenanceSchedule.from_json(schedule['maintenance']['schedule'])
+        schedule.validate()
+        return schedule
+
+    def put_schedule(self, schedule: MaintenanceSchedule):
```

Comment:
Needs return type annotation. I think `JSON` is satisfactory, but please verify that.

Reply:
The return value also isn't used anywhere, so you could remove it.